### DISCUSSION TITLE
Not able to draw lines using GeometryDecorator

### DISCUSSION THREAD

**Discussion by kardoka** [iTwin/itwinjs-core#4660] [Q&A] [answered]
Greetings,

I am trying to draw markers for XYZ coordinates coming in from WebSockets and lines for connecting these markers. For drawing the geometry I am using the `GeometryDecorator` from [one of the iTwin.js samples](https://www.itwinjs.org/sandboxes/iTwinPlatform/Simple%20Line). I create (and update) the decorator every 10 secs in `App.tsx` like this:

```
...

setInterval(() => {
        const oldDecorator = IModelApp.viewManager.decorators.find((dec) => dec instanceof GeometryDecorator);
        if (oldDecorator) {
          IModelApp.viewManager.dropDecorator(oldDecorator);
        }
        decorator = lineDrawer.getDecorator();
        IModelApp.viewManager.addDecorator(decorator);
      }, 10000);

...
```

In the static method `getDecorator()` I connect to a WebSockets server and retrieve the latest data and create the decorator. Then, inside of the WebSockets method `onmessage()`, I parse the input, create the markers and geometry, add them to the decorator:

```
...

public getDecorator() : GeometryDecorator {
    const decorator = new GeometryDecorator();
    decorator.clearGeometry();
    const ws = new WebSocket('ws://localhost:8999');
    ws.onmessage = (evt: MessageEvent) => { 
      const data: Point = JSON.parse(evt.data);
      this.points.push(data);
      this.points.forEach((point, i) => {
        var markerColor = ColorByName.gray;
        if(i === this.points.length - 1) {
          markerColor = ColorByName.green;
        }
        const marker = new InteractivePointMarker(
          Point3d.create(point.coordinates.x, point.coordinates.y, point.coordinates.z),
          `Time: ${new Date(point.time).toLocaleDateString("en-US", {
            hour: '2-digit',
            minute: '2-digit',
            second: '2-digit',
            hour12: false
          })}`,
          ColorDef.fromTbgr(ColorDef.withTransparency(ColorDef.create(markerColor).tbgr, 50)),
          () => { }
        );
        decorator.addMarker(marker); 
      });
      if(this.points.length > 1) {
        for(let i = 1; i <= this.points.length - 1; i++) {
          const pointA = Point3d.create(this.points[i - 1].coordinates.x, this.points[i - 1].coordinates.y, this.points[i - 1].coordinates.z);
          const pointB = Point3d.create(this.points[i].coordinates.x, this.points[i].coordinates.y, this.points[i].coordinates.z);
          const line = LineSegment3d.create(pointA, pointB);
          const geometry = {line: line, lineThickness: 2} as unknown as GeometryQuery;
          console.log(geometry);
          decorator.addGeometry(geometry);
        }
      }
      ws.close();
    };
    return decorator;
}

...
```

The markers for the retrieved XYZ coordinates render perfectly, but the lines are usually not visible, although they appear sometimes when I run the production build. Does anyone know more proper way of drawing these lines in my case?

**Comment by bbastings** [comment, marked as answer]
When you call IModelApp.viewManager.addDecorator it also calls ViewManager.invalidateDecorationsInAllViews to give the decorator a chance to regenerate it's graphics on the next render frame by calling it's "decorate" method. See [ViewportDecorator](https://www.itwinjs.org/reference/core-frontend/views/viewportdecorator/).

onmessage is likely being called after this initial "decorate" call for the GeometryDecorator, so you will need to explicitly call ViewManager.invalidateDecorationsInAllViews (or ScreenViewport.invalidateDecorators for specific views) yourself after adding geometry. You might want to consider just implementing your own Decorator and caching your line graphics instead of using GeometryDecorator.

**Reply by kardoka** [reply]
Thank you, I will try this out.

**Reply by kardoka** [reply]
@bbastings I tried invalidating the decorations right after adding the geometry, but the lines are still not drawn. What is interesting is that the markers are drawn perfectly all the time, although they are created and added to the same `GeometryDecorator` in the same `onmessage()` function as the lines.

```
...

decorator.addGeometry(geometry);
IModelApp.viewManager.invalidateDecorationsAllViews();

...
```

Regarding `onmessage()` it seems that the `GeometryDecorator.decorate()` function is always called after the geometry is added to the decorator inside `onmessage()`:

![image](https://user-images.githubusercontent.com/28572657/202150453-bfc773f8-7f7b-42e4-ba88-33b3caf89f07.png)

Do you have any tips on what I should try next?

**Reply by pmconne** [reply]
Try drawing the lines as overlay decorations. If they show up then, it means some other geometry is obscuring them.
If problems persist, show a screenshot of your view with markers and (possibly missing) lines.

**Reply by bbastings** [reply]
Are you specifying a color for your lines by calling setColor on the GeometryDecorator before calling addGeometry? It looks like the default line color used by GeometryDecorator is black...is your view backrground also black?

Why are you doing this?  const geometry = {line: line, lineThickness: 2} as unknown as GeometryQuery;

LineSegment3d is a subclass of GeometryQuery...

**Reply by kardoka** [reply]
Here is a screenshot of how it looks:

![image](https://user-images.githubusercontent.com/28572657/202201891-759868f6-1b94-450e-aeba-67ba2a0de90d.png)

Sometimes the black lines do flash, but only for a moment or so, like this:

![image](https://user-images.githubusercontent.com/28572657/202202132-ce109c04-dad6-490e-9d0d-549d0dbb513e.png)

> Try drawing the lines as overlay decorations. If they show up then, it means some other geometry is obscuring them. 

Thank you, I will try this out.

**Reply by kardoka** [reply]
> Are you specifying a color for your lines by calling setColor on the GeometryDecorator before calling addGeometry? It looks like the default line color used by GeometryDecorator is black...is your view backrground also black?

No, I am not setting a color for the lines. :) I have set the background to be white, so this should not be the issue as far as I know.

> Why are you doing this? const geometry = {line: line, lineThickness: 2} as unknown as GeometryQuery;
> 
> LineSegment3d is a subclass of GeometryQuery...

I was playing around with the line thickness, although I do recognize that this is not the most effective way of modifying it. Removing that line and just passing the `LineSegment3d` object to `GeometryDecorator.addGeometry()` does not seem to fix the issue though.

**Reply by kardoka** [reply]
@pmconne Thank you! It works perfectly now after using `GraphicBuilder` with the `GraphicType` of either `WorldDecoration` or `WorldOverlay`:

![image](https://user-images.githubusercontent.com/28572657/202399193-1e1c6177-c0bf-44cb-b9fb-a818c945b319.png)
